### PR TITLE
nxengine-evo: init at 2.6.4

### PR DESCRIPTION
**Pull request by scubed2**
###### Motivation for this change

Add a game.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

So far, I've done:
nix-env -i nxengine
to install it, then started a new game and played a bit.  For some reason, the music seems to be going fast, but that's the only issue I've encountered.  (So, only a cosmetic bug.  It is still fully playable, and the game itself is at the right speed.)

I see that the instructions say there should be a maintainer, but I don't know what to put there.

Note that the game has 2 parts:
the engine, which is opensource
the data, which is freeware and obtained separately
This package only has the engine.  So, it is not playable by itself.

### CODE REVIEW

File: pkgs/games/nxengine/default.nix
Comment (mmahut) [inline pkgs/games/nxengine/default.nix]: Please use ``pname``.

File: pkgs/games/nxengine/default.nix
Comment (mmahut) [inline pkgs/games/nxengine/default.nix]: Please use ``mirror://sourceforge/``.

File: pkgs/games/nxengine/default.nix
Comment (mmahut) [inline pkgs/games/nxengine/default.nix]: Please use quotes for the homepage if that's okay with you.

**Review by mmahut — CHANGES_REQUESTED** [review]
Fails for me after installation with:

```
$ nx 
Loading settings...
Couldn't open file settings.dat.
No saved settings; using defaults.
videoinfo: desktop bpp 32
Graphics::SetResolution(2)
Setting scaling 2 and fullscreen=no
SDL_SetVideoMode: 640x480 @ 32bpp
Graphics::FlushAll()
 error << SIFLoader::LoadHeader: failed to open file 'sprites.sif' >> 
 error << Failed to initialize graphics. >> 
```

**Comment by mmahut** [comment]
@GrahamcOfBorg build nxengine

File: pkgs/games/nxengine/default.nix
Comment (scubed2) [reply pkgs/games/nxengine/default.nix]: Replaced with pname.  Thanks!

File: pkgs/games/nxengine/default.nix
Comment (scubed2) [reply pkgs/games/nxengine/default.nix]: I'm not sure how that would work.  This is on the "site" page, not the releases.  So, I don't think mirror works here?  Or if it does, how do I change it to be of that form?

File: pkgs/games/nxengine/default.nix
Comment (scubed2) [reply pkgs/games/nxengine/default.nix]: Added quotes.
I don't have a good sense for when they're needed, since nix does seem to have URL literals.

**Comment by scubed2** [comment]
> Fails for me after installation with:

Yes.  This is intentional as written.  I looked around to see what other tools do.  I found archlinux does this:
https://aur.archlinux.org/cgit/aur.git/tree/nxengine.sh?h=nxengine
So, there's a wrapper script.  It downloads the freeware game, then sets up a directory, and cd's into there.  That way, it will be able to find all of its files, and will be sanely laid out.
I'm also not sure how mixing in the freeware would work legally.
So, I took the more conservative approach.

I can set it up with a script like that one, if so desired.  Then, it would start the game right away, which is convenient.

Please tell me your preference.

File: pkgs/games/nxengine/default.nix
Comment (mmahut) [reply pkgs/games/nxengine/default.nix]: Oh you are right, I suspected the link would be to releases. Odd.

**Comment by mmahut** [comment]
@scubed2 I think it would be nice to have this working out of box. So a kind of script/wrapper to do that, plus moving the package to unfree if required.

**Comment by aanderse** [comment]
It wasn't immediately obviously to me where to get the data from when looking at the homepage. Is the data unfree **and** not redistributable?

**Comment by scubed2** [comment]
Okay, the package should be completely working now.

mmahut:

* the game assets are now included in share
* the license is changed to be gpl3, unfreeRedistributable to account for having the assets
* instead of a wrapper, I sed $out into the binary
  * the assets are released with the binary version, so this seems reasonable to me
* music plays at the correct speed

aanderse:

The game assets are redistributable.
https://www.cavestory.org/forums/threads/cave-story-freeware-license.13817/
Note that this is Cave Story, not to be confused with Cave Story+, which is fully proprietary and non-redistributable.

**Comment by scubed2** [comment]
For reference, the patches have been sent as:
https://github.com/nxengine/nxengine-evo/pull/150
https://github.com/nxengine/nxengine-evo/pull/151

**Comment by aanderse** [comment]
@scubed2 I see your changes have been merged upstream. Great work!
Please fetch the patches from upstream so we don't need to maintain them in NixOS.

**Comment by scubed2** [comment]
I've downloaded the patches from upstream and put them into pixtone.patch.  So, now it matches.

Either way, the patches wouldn't need to be maintained, since the next release would make both patches obsolete.

I also recompiled and ran it for a few minutes to sanity check.

**Comment by aanderse** [comment]
@scubed2 I was referring to using `fetchpatch` so we directly reference the patch upstream instead of keeping a local copy in `nixpkgs`. If you want to update your PR to use `fetchpatch` we can merge now, or if there is no rush just wait for the next release and update.

**Comment by scubed2** [comment]
I've changed it to use fetchpatch from the github commits of the upstream changes.  So, hopefully it should be good now.

For my own curiosity, how different is it to have the patch locally vs. using fetchpatch?

**Review by aanderse — CHANGES_REQUESTED** [review]
We don't like to maintain patches in `nixpkgs` if they exist elsewhere and can be pulled down. Duplication of code, among other things.

I ran the game and had a few errors like this: 
```
error << pxt->load: file '/nix/store/gds3sl3alkqwzzd9wk47xywldvy224lb-nxengine-evo-2.6.4/share/nxengine/data/pxt/fx08.pxt' not found. >>
```
The game seemed booted up, so I'm not sure if that is a problem or not.

**Comment by scubed2** [comment]
I've switched to using nativeBuildInputs.  It still builds and runs.

The pxt files are the sound effects.  If some are missing, then it just won't play that sound effect (that is, it will be silent).  So, missing some would be innocuous.  I've played the game and haven't noticed anything missing.  So, it seems to be working.  I'm not sure how to be certain if sound effect 8 is used.  That is, at worst it will cause some sound effects to be missing, but likely it is working as intended.

**Comment by aanderse** [comment]
@scubed2 You should add yourself to `maintainers/maintainer-list.nix` as the first commit to this PR. Then you can amend your second commit to include yourself in the `meta.maintainers` field.

Aside from that, we're good to go. Thanks!

**Comment by scubed2** [comment]
Added myself as a maintainer in a separate commit.

**Comment by aanderse** [comment]
@GrahamcOfBorg build nxengine-evo

**Pull request merged**
